Feature/split dataset api tests

nshumoogum opened the pull request:
### What

Split out dataset api tests into web and publishing to be run against web and publishing subnets respectively.

### How to review

Check tests pass against `feature/observation-spec` branch on `dp-dataset-api`, remember to run `make acceptance-web` or `make acceptance-publishing` to start service in web or publishing respectively. You will also need `dp-auth-api-stub` running for publishing tests to pass.

### Who can review

Team B

nshumoogum committed:
Add dataset API web subnet tests for endpoints hidden in web

Tests cover check that endpoint returns a status of not found (404) even
if the request has a valid authentication header.

nshumoogum committed:
Move and update dataset api tests to web tests

nshumoogum committed:
Add dataset API tests that are only seen on publishing subnet

nshumoogum committed:
Move dataset API tests to publishing ones

Non GET requests previously moved over and updated, so delete those ones
from datasetAPI folder in root directory.

nshumoogum committed:
Remove global constant from web/datasetAPI package on initialisation

nshumoogum committed:
Update datasetAPI README.md for publishing and web

nshumoogum committed:
Add new test to check multi-value query parameters

... for dataset API observation endpoint

nshumoogum committed:
Extend e2e test to call observations endpoint...

...once data is in neo4j for pre-publishing and again once the version
is published without an auth header

nshumoogum committed:
Update response error message for observations endpoint

eldeal commented on publishing/datasetAPI/put_instance_test.go:
```diff
@@ -59,9 +62,8 @@ func TestSuccessfullyPutInstance(t *testing.T) {
-			datasetAPI.PUT("/instances/{instance_id}", instanceID).
-				WithHeader(florenceTokenName, florenceToken).WithBytes([]byte(validPUTCompletedInstanceJSON)).
-				Expect().Status(http.StatusOK)
+			datasetAPI.PUT("/instances/{instance_id}", instanceID).WithHeader(florenceTokenName, florenceToken).
```
It seems inconsistent that there's a few cases in this file which are only across 2 lines instead of 4/5? Some of the lines are longer than ones that have been split out in other cases

eldeal commented on publishing/datasetAPI/put_version_test.go:
```diff
@@ -440,19 +464,10 @@ func TestFailureToUpdateVersion(t *testing.T) {
-		})
-
-		// test for bad request when publishing version (Missing mandatory fields)
-		Convey("When an authorised PUT request is made to update version resource to a state of published", func() {
```
It feels like we probably still need some test to assert that a document can't go straight from edition confirmed to publish without meeting certain prereqs (at least the same as the associated) - though I haven't looked to see if one is buried somewhere else in the file admittedly

nshumoogum replied:
As discussed earlier, the `collection_id` is now removed on publishing. Validation of field could be done on earlier states of the resource. None the less I have added some tests to check other mandatory fields are set before publishing and invalid ones too.

eldeal commented on web/datasetAPI/get_datasets_test.go:
```diff
@@ -0,0 +1,153 @@
+		Collection: "datasets",
+		Key:        "_id",
+		Value:      "133",
+		Update:     validAssociatedDatasetData(datasetID),
```
Should the update be initialized with the actual dataset ID for this document - in this case 133?

nshumoogum replied:
yes, i've made this a variable to be reused and is not hardcoded to `133`

eldeal commented on web/datasetAPI/get_datasets_test.go:
```diff
@@ -0,0 +1,153 @@
+
+					if response.Value("items").Array().Element(i).Object().Value("id").String().Raw() == datasetID {
+						// check the published test dataset document has the expected returned fields and values
+						response.Value("items").Array().Element(i).Object().Value("id").Equal(datasetID)
```
isnt this functionally the same check as the if statement? probably dont need both

eldeal commented on web/datasetAPI/get_datasets_test.go:
```diff
@@ -0,0 +1,153 @@
+						checkDatasetResponse(datasetID, response.Value("items").Array().Element(i).Object())
+					}
+
+					if response.Value("items").Array().Element(i).Object().Value("id").String().Raw() == "133" {
```
again, this feels like the same assertion from further up this test, seems redundant

eldeal commented on web/datasetAPI/get_edition_test.go:
```diff
@@ -0,0 +1,181 @@
+	datasetAPI := httpexpect.New(t, cfg.DatasetAPIURL)
+
+	Convey("Given a published edition of a dataset", t, func() {
+		Convey("When a GET requestis made to retrieve the edition", func() {
```
space in requestis please :)

eldeal commented on web/datasetAPI/get_editions_test.go:
```diff
@@ -0,0 +1,155 @@
+		Database:   cfg.MongoDB,
+		Collection: "editions",
+		Key:        "_id",
+		Value:      "466",
```
This value isnt passed in to the update - should it be?

eldeal commented on web/datasetAPI/get_version_test.go:
```diff
@@ -0,0 +1,270 @@
+	}
+}
+
+func setupPublishedAndUnpublishedVersions(datasetID, editionID, edition, instanceID string) ([]*mongo.Doc, error) {
```
this function name seems misleading - this only setups published documents

eldeal reviewed (changes requested):
The readmes mention a make target but I can't find any Makefiles?

nshumoogum committed:
Check for mandatory fields when publishing version

* Also includes fix to typos
* When iterating list/items check the number of items inspected;
 this is because if the item is not found in the list and inspection of
 the item will not take place and hence the test would not have failed
* Remove unnecessary checks on response data, (already handled by an if
 statement)

nshumoogum committed:
Update publishing subnet tests

If an auth header is not set on request, api will always respond with
unauthorized (401) status (code).

nshumoogum commented:
@eldeal  re. make targets, this is referring to the application makefile. And now that everything is merged together in `feature/observation-spec` branch on the dataset API you should be able to find them there and be able to run the service on that branch in the dataset API for the acceptance tests to pass.

Don't forget to run `dp-auth-api-stub` too for publishing tests

eldeal commented on publishing/datasetAPI/get_dataset_test.go:
```diff
@@ -17,6 +17,7 @@ import (
 func TestSuccessfullyGetADataset(t *testing.T) {
 
 	datasetID := uuid.NewV4().String()
+	unpublishedDatasteID := uuid.NewV4().String()
```
there is a typo in the variable name

eldeal commented on publishing/datasetAPI/get_editions_test.go:
```diff
@@ -117,10 +107,10 @@ func TestFailureToGetListOfDatasetEditions(t *testing.T) {
 	Convey("Given the dataset does not exist", t, func() {
 		Convey("When a request to get editions for a dataset is made", func() {
-			Convey("Then return a status not found (404)", func() {
+			Convey("Then return a status unauthorized (404)", func() {
```
unauthorized is not a 404 :)

eldeal commented on publishing/datasetAPI/put_version_test.go:
```diff
@@ -436,15 +440,75 @@ func TestFailureToUpdateVersion(t *testing.T) {
 	})
 
+	// test for updating a version so it with a state of `published` but missing mandatory fields
+	Convey("Given a published dataset and edition and an unpublished version exist ...", t, func() {
```
dont need `...` in the convey statements throughout this test. also `so it with a state` does not make sense in the comment

eldeal commented on publishing/datasetAPI/put_version_test.go:
```diff
@@ -436,15 +440,75 @@ func TestFailureToUpdateVersion(t *testing.T) {
+			}
+
+			Convey("When an authorised PUT request is made to update version resource to published", func() {
+				Convey("Then fail to update resource and return a status bad request (400) with a message `Missing collection_id for association between version and a collection`", func() {
```
it doesnt feel like the end of this message matches what is being asserted in Contains() - this could just say 'with the correct message' or it should test for both aspects of the outcome if theres genuinely two messages (same below)

eldeal commented on publishing/datasetAPI/get_datasets_test.go:
```diff
@@ -48,32 +49,13 @@ func TestSuccessfulGetAListOfDatasets(t *testing.T) {
-				}
-			}
-		})
+			expectedDatasets := make(map[string]int)
```
why is this a map and not just a count? the same incrementing could happen in the for loop

nshumoogum replied:
So firstly yes it could just be a count of int type and secondly we wouldn't want it in the for loop as some people do not use the test database to set up test data when running tests, and hence they could have more than two datasets being returned, hence we need to look for the two datasets that we are inspecting/checking. 

Also if someone does not teardown there test data between scenarios correctly this could give a false negative here.

I will change variable to an int however

nshumoogum committed:
Fix typos

nshumoogum committed:
Update dataset API tests due to changes in test data on setup

nshumoogum committed:
Merge branch 'cmd-develop' into feature/split-dataset-api-tests

nshumoogum committed:
Fix typo
